### PR TITLE
std/crypto: vectorize Salsa20

### PR DESCRIPTION
**Pull request by jedisct1**
20% faster on x86_64, slower on aarch64 as usual :/

### CODE REVIEW

File: lib/std/crypto/salsa20.zig [hunk 1]
```diff
@@ -9,11 +9,185 @@ const crypto = std.crypto;
...
+            Lane{ constant_le[0], constant_le[1], constant_le[2], constant_le[3] },
+            Lane{ d[0], d[1], d[2], d[3] },
+        };
+    }
+
+    inline fn rot(x: Lane, comptime n: u5) Lane {
+        return (x << @splat(4, @as(u5, n))) | (x >> @splat(4, @as(u5, 1 +% ~n)));
+    }
+
+    inline fn salsa20Core(x: *BlockVec, input: BlockVec, comptime feedback: bool) void {
```
Comment (daurnimator) [inline lib/std/crypto/salsa20.zig:43]: I feel like this `inline` is going to bloat out small builds.
Does it really not inline itself?

Comment (jedisct1) [reply lib/std/crypto/salsa20.zig:43]: An explicit `inline` makes a slight difference (`ReleaseFast`):

no inline:
```text
 xsalsa20Poly1305:        554 MiB/s
```

inline:
```text
 xsalsa20Poly1305:        569 MiB/s
```

With `ReleaseSmall`, the size doesn't change at all with and without.

**Comment by LemonBoy** [comment]
DJB wrote a [paper](https://cr.yp.to/papers.html#neoncrypto) about Neon-accelerated Salsa20 implementations, it may be useful as a reference to find out why the vector code performs so badly on aarch64.

**Comment by jedisct1** [comment]
The exact same operations implemented using C intrinsics are actually quite fast. That was also the case for other primitives.

Suboptimal scheduling when using generic vector instructions maybe?

NEON (and more generally non-x86_64) optimizations is something I will take a closer look at after the `0.7` release. The entire country is going to be locked down starting tonight, so I may have some spare time to spend during the forthcoming weeks (months?).

**Comment by andrewrk** [comment]
> The entire country is going to be locked down starting tonight, so I may have some spare time to spend during the forthcoming weeks (months?).

I'm sorry to hear that, hope you make it though OK.

> The exact same operations implemented using C intrinsics are actually quite fast. That was also the case for other primitives.

This is a useful case study. Perhaps this is something that could be improved in LLVM's aarch64 vector codegen? I reckon we will gain some insight into this when we write the codegen for x86_64/aarch64 vector operations in stage2.

**Pull request merged**